selftests/unit/test_vm.py

```diff
@@ -1,71 +1,66 @@
-        mock_vm.should_receive('create_snapshot').once().ordered()
-        # VMTestRunner()
-        Args = flexmock(test_result_total=1,
-                        url=['/tests/sleeptest', '/tests/other/test',
-                             'passtest'],
-                        vm_domain='domain',
-                        vm_username='username',
-                        vm_hostname='hostname',
-                        vm_port=22,
-                        vm_password='password',
-                        vm_key_file=None,
-                        vm_cleanup=True,
-                        vm_no_copy=False,
-                        vm_timeout=120,
-                        vm_hypervisor_uri='my_hypervisor_uri',
-                        env_keep=None)
-        log = flexmock()
-        log.should_receive("info")
-        job = flexmock(args=Args, log=log)
-        self.runner = avocado_runner_vm.VMTestRunner(job, None)
-        mock_vm.should_receive('stop').once().ordered()
-        mock_vm.should_receive('restore_snapshot').once().ordered()
-
-    def tearDown(self):
-        flexmock_teardown()
-
     def test_setup(self):
-        """ Tests VMTestRunner.setup() """
-        self.runner.setup()
-        self.runner.tear_down()
-        flexmock_teardown()
+        mock_vm = _FakeVM()
+        mock_vm.start = mock.Mock(return_value=True)
+        mock_vm.create_snapshot = mock.Mock()
+        mock_vm.stop = mock.Mock()
+        mock_vm.restore_snapshot = mock.Mock()
+        job_args = argparse.Namespace(test_result_total=1,
+                                      vm_domain='domain',
+                                      vm_username='username',
+                                      vm_hostname='hostname',
+                                      vm_port=22,
+                                      vm_password='password',
+                                      vm_key_file=None,
+                                      vm_cleanup=True,
+                                      vm_no_copy=False,
+                                      vm_timeout=120,
+                                      vm_hypervisor_uri='my_hypervisor_uri',
+                                      reference=['/tests/sleeptest.py',
+                                                 '/tests/other/test',
+                                                 'passtest.py'],
+                                      dry_run=True,
+                                      env_keep=None)
+        with mock.patch('avocado_runner_vm.vm_connect', return_value=mock_vm):
+            # VMTestRunner()
+            job = Job(job_args)
+            runner = avocado_runner_vm.VMTestRunner(job, None)
+            runner.setup()
+            runner.tear_down()
+        mock_vm.start.assert_called_once()
+        mock_vm.create_snapshot.assert_called_once()
+        mock_vm.stop.assert_called_once()
+        mock_vm.restore_snapshot.assert_called_once()
+        shutil.rmtree(job.args.base_logdir)
```

Comment:
It'd be safer to provide `base_logdir` (in case `Job` creation fails). Also you ought to clean it even on failure (at least try/finally, ideally setUp/tearDown)

Reply:
A try/finally block, for a single test, is as good as setUp/tearDown for classes, so I'll go with that.